File: src/main/java/no/hvl/tk/visual/debugger/server/VisualDebuggingAPIServerStarter.java
package no.hvl.tk.visual.debugger.server;

import com.intellij.openapi.diagnostic.Logger;
import jakarta.websocket.Session;
import no.hvl.tk.visual.debugger.server.endpoint.UIConfig;
import no.hvl.tk.visual.debugger.server.endpoint.VisualDebuggingAPIEndpoint;
import no.hvl.tk.visual.debugger.server.endpoint.message.DebuggingWSMessage;
import no.hvl.tk.visual.debugger.server.endpoint.message.DebuggingMessageType;
import no.hvl.tk.visual.debugger.settings.PluginSettingsState;
import org.glassfish.tyrus.server.Server;

import java.io.IOException;
import java.util.HashMap;

/**
 * This class can start a websocket server which runs an API which provides the client with live debug data.
 * See {@link VisualDebuggingAPIEndpoint} for the Endpoint.
 */
public class VisualDebuggingAPIServerStarter {
    private static final Logger LOGGER = Logger.getInstance(VisualDebuggingAPIServerStarter.class);

    private VisualDebuggingAPIServerStarter() {
        // Only helper methods.
    }

    public static Server runNewServer() {
        final Server server = new Server(
                ServerConstants.HOST_NAME,
                ServerConstants.VISUAL_DEBUGGING_API_SERVER_PORT,
                "",
                new HashMap<>(),
                VisualDebuggingAPIEndpoint.class);
        try {
            server.start();
            LOGGER.info("Debug API server started successfully.");
            return server;
        } catch (final Exception e) {
            LOGGER.error(e);
            return null;
        }
    }

    /**
     * Sends the given message to the given client, if the client is not null.
     *
     * @param client  client.
     * @param message message for the client.
     */
    public static void sendMessageToClient(final Session client, final String message) {
        if (client != null) {
            try {
                client.getBasicRemote().sendText(message);
            } catch (final IOException e) {
                LOGGER.error(e);
            }
        }
    }

    /**
     * Sends the configuration to the client.
     * @param client client
     */
    public static void sendUIConfig(Session client) {
        UIConfig uiConfig = PluginSettingsState.getInstance().getUIConfig();
        final DebuggingWSMessage configMessage = new DebuggingWSMessage(
                DebuggingMessageType.CONFIG,
                uiConfig.serialize());

        sendMessageToClient(client, configMessage.serialize());
    }

}

File: src/main/java/no/hvl/tk/visual/debugger/server/VisualDebuggingAPIServerStarter.kt
package no.hvl.tk.visual.debugger.server

import com.intellij.openapi.diagnostic.Logger
import jakarta.websocket.Session
import java.io.IOException
import no.hvl.tk.visual.debugger.server.endpoint.VisualDebuggingAPIEndpoint
import no.hvl.tk.visual.debugger.server.endpoint.message.DebuggingMessageType
import no.hvl.tk.visual.debugger.server.endpoint.message.DebuggingWSMessage
import no.hvl.tk.visual.debugger.settings.PluginSettingsState.Companion.settings
import org.glassfish.tyrus.server.Server

/**
 * This class can start a websocket server which runs an API that provides the client with live
 * debug data. See [VisualDebuggingAPIEndpoint] for the Endpoint.
 */
object VisualDebuggingAPIServerStarter {
  private val LOGGER = Logger.getInstance(VisualDebuggingAPIServerStarter::class.java)

  @JvmStatic
  fun runNewServer(): Server? {
    val server =
        Server(
            ServerConstants.HOST_NAME,
            ServerConstants.VISUAL_DEBUGGING_API_SERVER_PORT,
            "",
            HashMap(),
            VisualDebuggingAPIEndpoint::class.java)
    try {
      server.start()
      LOGGER.info("Debug API server started successfully.")
      return server
    } catch (e: Exception) {
      LOGGER.error(e)
      return null
    }
  }

  /**
   * Sends the given message to the given client, if the client is not null.
   *
   * @param client client.
   * @param message message for the client.
   */
  @JvmStatic
  fun sendMessageToClient(client: Session, message: String) {
      try {
        client.basicRemote.sendText(message)
      } catch (e: IOException) {
        LOGGER.error(e)
      }
  }

  /**
   * Sends the configuration to the client.
   *
   * @param client client
   */
  @JvmStatic
  fun sendUIConfig(client: Session) {
    val uiConfig = settings.uIConfig
    val configMessage = DebuggingWSMessage(DebuggingMessageType.CONFIG, uiConfig.serialize())

    sendMessageToClient(client, configMessage.serialize())
  }
}
